engine-precompiles/src/async_router.rs

```diff
@@ -0,0 +1,105 @@
+use super::{EvmPrecompileResult, Precompile};
+use crate::prelude::types::{Address, EthGas};
+use crate::prelude::{vec, Vec};
+use crate::PrecompileOutput;
+use aurora_engine_sdk::types::near_account_to_evm_address;
+use aurora_engine_types::account_id::AccountId;
+use aurora_engine_types::parameters::{CallArgs, FunctionCallArgsV2};
+use aurora_engine_types::types::WeiU256;
+use aurora_engine_types::Cow;
+use borsh::BorshSerialize;
+use evm::backend::Log;
+use evm::{Context, ExitError};
+
+const ERR_INVALID_INPUT: &str = "ERR_INVALID_ROUTER_INPUT";
+mod costs {
+    use crate::prelude::types::EthGas;
+
+    // TODO(#483): Determine the correct amount of gas
+    pub(super) const ASYNC_ROUTER_GAS: EthGas = EthGas::new(0);
+}
+
+pub struct AsyncRouter;
+
+/// async_router precompile address
+///
+/// Address: `0xad65a767211ae644cdf2d036853e2bcda225dff8`
+/// This address is computed as: `&keccak("asyncRouter")[12..]`
+pub const ADDRESS: Address = crate::make_address(0xad65a767, 0x211ae644cdf2d036853e2bcda225dff8);
+
+fn predecessor_address(predecessor_account_id: &AccountId) -> Address {
+    near_account_to_evm_address(predecessor_account_id.as_bytes())
+}
+
+impl Precompile for AsyncRouter {
```

Comment:
Another idea would be to implement `AsyncRouter` as a regular Solidity contract and create a precompile that will call that router contract. The router contract could be deployed using a separate function (e.g. `deploy_router_contract()`.

Reply:
Yes, this is the point of contention, I also preferred the regular solidity contract, but I got request to implement it as precompile